<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Исследование-баланса-классов" data-toc-modified-id="Исследование-баланса-классов-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Исследование баланса классов</a></span></li><li><span><a href="#Построение-моделей-без-учета-дисбаланса" data-toc-modified-id="Построение-моделей-без-учета-дисбаланса-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Построение моделей без учета дисбаланса</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li></ul></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#RandomOverSampler" data-toc-modified-id="RandomOverSampler-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>RandomOverSampler</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li></ul></li><li><span><a href="#SMOTE" data-toc-modified-id="SMOTE-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>SMOTE</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li></ul></li><li><span><a href="#ADASYN" data-toc-modified-id="ADASYN-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>ADASYN</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.3.2"><span class="toc-item-num">3.3.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.3.3"><span class="toc-item-num">3.3.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li></ul></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.Hам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

**Цель** - необходимо построить модель, которая будет прогнозировать, уйдёт клиент из банка в ближайшее время или нет. 

В качестве метрики качества необходимо использовать *F1*-меру. По тех.заданию нужно получить метрику не менее 0.59. 

Дополнительно необходимо измерить *AUC-ROC*, сравнить её значение с *F1*-мерой.

## Подготовка данных

In [1]:
# загружаем необходимые библиотеки
!pip install imbalanced-learn
!pip install skimpy

import pandas as pd
import matplotlib.pyplot as plt

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, roc_curve, roc_auc_score
from sklearn.model_selection import train_test_split
from sklearn.utils import shuffle
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.model_selection import GridSearchCV

from collections import Counter

from imblearn.over_sampling import RandomOverSampler, ADASYN, SMOTE

from skimpy import clean_columns

In [2]:
#загружаем датасет
try:
    df = pd.read_csv('/Users/amirk/Downloads/Churn.csv')
    df.head()
except:
    df = pd.read_csv('/datasets/Churn.csv')
    df.head()
    

In [3]:
# поправим названия столбцов
df = clean_columns(df)
df.head()

14 column names have been cleaned

row_number  customer_id   surname  credit_score geography  gender  age  \
0           1     15634602  Hargrave           619    France  Female   42   
1           2     15647311      Hill           608     Spain  Female   41   
2           3     15619304      Onio           502    France  Female   42   
3           4     15701354      Boni           699    France  Female   39   
4           5     15737888  Mitchell           850     Spain  Female   43   

   tenure    balance  num_of_products  has_cr_card  is_active_member  \
0     2.0       0.00                1            1                 1   
1     1.0   83807.86                1            0                 1   
2     8.0  159660.80                3            1                 0   
3     1.0       0.00                2            0                 0   
4     2.0  125510.82                1            1                 1   

   estimated_salary  exited  
0         101348.88       1  
1         112542.58       0  
2         113931.57       1  
3          93826.63       0  
4          79084.10       0

In [4]:
# познакомимся с данными
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   row_number        10000 non-null  int64  
 1   customer_id       10000 non-null  int64  
 2   surname           10000 non-null  object 
 3   credit_score      10000 non-null  int64  
 4   geography         10000 non-null  object 
 5   gender            10000 non-null  object 
 6   age               10000 non-null  int64  
 7   tenure            9091 non-null   float64
 8   balance           10000 non-null  float64
 9   num_of_products   10000 non-null  int64  
 10  has_cr_card       10000 non-null  int64  
 11  is_active_member  10000 non-null  int64  
 12  estimated_salary  10000 non-null  float64
 13  exited            10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


- Данные распределены достаточно правдоподобно. В столбце `tenure` выявлено 909 пропусков, это 9% от всей выборки, удалим данные строки, так-так нет уверенности, что там были нули, например.     
- В столбцах `has_credit_card` и `is_active_member` бинарные категориальные данные, поэтому дополнительная кодировка данных не требуется.
- Дубликатов в данных нет
- Строки `row_number`, `customer_id` и `surname` не несут в себе важной информации для построения моделей, удалим их.
- Строки `gender` и `geography` имеют тип данных **object**, выполним кодировку данных столбцов
- Выополним масштабирование количественных признаков в столбцах `credit_score`, `age`, `tenure`, `balance`, `num_of_products`, `estimated_salary` после разделения на тренировочную и тестовую выборки.


In [5]:
# удаляем ненужные строки
df_final = df.drop(['row_number', 'customer_id', 'surname'], axis=1)

# кодируем данные
obj_columns = df_final.select_dtypes('object').columns

ohe = OneHotEncoder(sparse=False, drop='first')
df_ohe = ohe.fit_transform(df_final[obj_columns])
df_final[ohe.get_feature_names_out()] = pd.DataFrame(df_ohe, columns=ohe.get_feature_names_out())
df_final.drop(obj_columns, axis=1, inplace=True)

#удаляем пропуски в tenure
df_final.dropna(inplace=True)
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9091 entries, 0 to 9998
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   credit_score       9091 non-null   int64  
 1   age                9091 non-null   int64  
 2   tenure             9091 non-null   float64
 3   balance            9091 non-null   float64
 4   num_of_products    9091 non-null   int64  
 5   has_cr_card        9091 non-null   int64  
 6   is_active_member   9091 non-null   int64  
 7   estimated_salary   9091 non-null   float64
 8   exited             9091 non-null   int64  
 9   geography_Germany  9091 non-null   float64
 10  geography_Spain    9091 non-null   float64
 11  gender_Male        9091 non-null   float64
dtypes: float64(6), int64(6)
memory usage: 923.3 KB


In [6]:
# разделим данные на параметры и целевой признак, а затем на тренировочную и тестовую выборки
 
X = df_final.drop('exited', axis=1)
y = df_final['exited']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=123)

print(X.shape, y.shape)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(9091, 11) (9091,)
(6818, 11) (2273, 11) (6818,) (2273,)


In [7]:
# масштабирем данные
num_columns = ['credit_score', 'age', 'tenure', 'balance', 'num_of_products', 'estimated_salary']
scaler = StandardScaler()
X_train[num_columns] = scaler.fit_transform(X_train[num_columns])
X_test[num_columns] = scaler.transform(X_test[num_columns])
display(X_train.head())
display(X_test.head())

credit_score       age    tenure   balance  num_of_products  \
1529     -0.758136 -1.505822  1.035484 -1.229833         0.788837   
2194      2.076359 -0.469039 -1.033562 -1.229833         0.788837   
4881     -0.747715 -0.751798  0.345802 -1.229833         0.788837   
5482     -0.497613  0.661997 -1.033562 -1.229833         0.788837   
7853      0.825846  0.756250  0.690643 -1.229833        -0.914771   

      has_cr_card  is_active_member  estimated_salary  geography_Germany  \
1529            1                 0          0.193903                0.0   
2194            1                 1          1.231263                0.0   
4881            1                 0         -1.302997                0.0   
5482            1                 0          1.279524                0.0   
7853            1                 0         -1.177222                0.0   

      geography_Spain  gender_Male  
1529              1.0          1.0  
2194              0.0          0.0  
4881              0.0          1.0  
5482              1.0          1.0  
7853              0.0          0.0

credit_score       age    tenure   balance  num_of_products  \
7176     -1.623074  0.002226  1.035484  0.696527         0.788837   
5276     -0.789399 -0.940304 -0.343880  0.721700         0.788837   
2281      2.055517  0.096479  0.000961  1.148972        -0.914771   
3054     -1.154132  2.641310 -1.378403  0.185990        -0.914771   
9692     -0.799820  1.227515  0.690643  0.620924        -0.914771   

      has_cr_card  is_active_member  estimated_salary  geography_Germany  \
7176            1                 1         -1.581378                1.0   
5276            1                 1         -0.881316                0.0   
2281            0                 0          1.007489                1.0   
3054            0                 1         -0.723085                0.0   
9692            1                 0          1.658714                0.0   

      geography_Spain  gender_Male  
7176              0.0          1.0  
5276              0.0          1.0  
2281              0.0          0.0  
3054              1.0          1.0  
9692              1.0          1.0

## Исследование задачи

### Исследование баланса классов

In [8]:
sorted(Counter(y).items())

[(0, 7237), (1, 1854)]

В данных имеется дисбаланс классов.

### Построение моделей без учета дисбаланса

#### Дерево решений

In [9]:
def best_model_dtc(X_train, y_train):
    model = DecisionTreeClassifier(random_state=123)
    params = {'max_depth':range(1,20), 
              'min_samples_split':range(2,10), 
              'min_samples_leaf':range(1,10)}
    grid_model = GridSearchCV(model, 
                              params, 
                              cv=5, 
                              scoring=['f1', 'roc_auc'], 
                              refit='f1', 
                              error_score='raise', 
                              n_jobs=-1, 
                              verbose=1)
    grid_model.fit(X_train, y_train)
    print('Значение F1-score = ', grid_model.best_score_)
    print('Значение ROC-AUC = ', grid_model.cv_results_['mean_test_roc_auc'].mean())
    print('Лучшие параметры модели:', grid_model.best_params_)
    return grid_model

In [10]:
best_model_dtc_dis = best_model_dtc(X_train, y_train)

Fitting 5 folds for each of 1368 candidates, totalling 6840 fits
Значение F1-score =  0.5558629920157794
Значение ROC-AUC =  0.7691352279843282
Лучшие параметры модели: {'max_depth': 7, 'min_samples_leaf': 7, 'min_samples_split': 2}


#### Случайный лес

In [11]:
# Для начала найдем лучшие параметры модели на небольшом количестве деревьев 
#с целью экономии времени обучения, а в целом можно и нужно количество деревьев сразу смотреть.
# Затем подберем такой гиперпараметр как количество деревьев
def best_model_rfc(X_train, y_train):
    model = RandomForestClassifier(random_state=123, n_estimators=20)
    params = {'max_depth':range(1,20), 
              'min_samples_split':range(2,10), 
              'min_samples_leaf':range(1,10)}
    grid_model = GridSearchCV(model, 
                              params, 
                              cv=5, 
                              scoring=['f1', 'roc_auc'], 
                              refit='f1', 
                              error_score='raise', 
                              n_jobs=-1, 
                              verbose=1)
    grid_model.fit(X_train, y_train)
    model = RandomForestClassifier(random_state=123)
    params = {'n_estimators':range(20,350,10), 
              'max_depth':[grid_model.best_params_['max_depth']],
              'min_samples_split':[grid_model.best_params_['min_samples_split']],
              'min_samples_leaf':[grid_model.best_params_['min_samples_leaf']]}
    grid_model = GridSearchCV(model, 
                              params, 
                              cv=5, 
                              scoring=['f1', 'roc_auc'], 
                              refit='f1', 
                              error_score='raise', 
                              n_jobs=-1, 
                              verbose=1)
    grid_model.fit(X_train, y_train)
    print('Значение F1-score = ', grid_model.best_score_)
    print('Значение ROC-AUC = ', grid_model.cv_results_['mean_test_roc_auc'].mean())
    print('Лучшие параметры модели:', grid_model.best_params_)
    return grid_model

In [12]:
best_model_rfc_dis = best_model_rfc(X_train, y_train)

Fitting 5 folds for each of 1368 candidates, totalling 6840 fits
Fitting 5 folds for each of 33 candidates, totalling 165 fits
Значение F1-score =  0.5804889396696881
Значение ROC-AUC =  0.8557447193252888
Лучшие параметры модели: {'max_depth': 18, 'min_samples_leaf': 3, 'min_samples_split': 2, 'n_estimators': 60}


#### Логистическая регрессия

In [13]:
def best_model_lr(X_train, y_train):
    model = LogisticRegression(random_state=123, class_weight='balanced')
    params = {'solver':['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga'], 
              'max_iter':range(100,1000,100), 
              'C':[1, 10, 100, 1000]}
    grid_model = GridSearchCV(model, 
                              params, 
                              cv=5, 
                              scoring=['f1', 'roc_auc'], 
                              refit='f1', 
                              error_score='raise', 
                              n_jobs=-1, 
                              verbose=1)
    grid_model.fit(X_train, y_train)
    print('Значение F1-score = ', grid_model.best_score_)
    print('Значение ROC-AUC = ', grid_model.cv_results_['mean_test_roc_auc'].mean())
    print('Лучшие параметры модели:', grid_model.best_params_)
    return grid_model 

In [14]:
best_madel_lr_dis = best_model_lr(X_train, y_train)

Fitting 5 folds for each of 180 candidates, totalling 900 fits
Значение F1-score =  0.4977726603059809
Значение ROC-AUC =  0.7709287338988858
Лучшие параметры модели: {'C': 10, 'max_iter': 100, 'solver': 'newton-cg'}


Без учета баланса классов лучшей моделью прогнозирования стала модель типа `RandomForest`. Модель имела следующие параметры: 
- **max_depth = 18**,
- **min_samples_leaf = 3**,
- **min_samples_split = 2**,
- **n_estimators = 60**

**F1 score = 0.580**   
**ROC_AUC = 0.856**

## Борьба с дисбалансом

### RandomOverSampler

In [15]:
X_ros, y_ros = RandomOverSampler(random_state=123).fit_resample(X_train, y_train)

print("Original dataset:", sorted(Counter(y_train).items()))

print("ROS-sampled dataset:", sorted(Counter(y_ros).items()))

Original dataset: [(0, 5427), (1, 1391)]
ROS-sampled dataset: [(0, 5427), (1, 5427)]


#### Дерево решений 

In [16]:
best_model_dtc_ros = best_model_dtc(X_ros, y_ros)

Fitting 5 folds for each of 1368 candidates, totalling 6840 fits
Значение F1-score =  0.9234786524181084
Значение ROC-AUC =  0.8692277853316248
Лучшие параметры модели: {'max_depth': 19, 'min_samples_leaf': 1, 'min_samples_split': 2}


#### Случайный лес

In [17]:
best_model_rfc_ros = best_model_rfc(X_ros, y_ros)

Fitting 5 folds for each of 1368 candidates, totalling 6840 fits
Fitting 5 folds for each of 33 candidates, totalling 165 fits
Значение F1-score =  0.9576403295596382
Значение ROC-AUC =  0.9946876598431369
Лучшие параметры модели: {'max_depth': 19, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 290}


#### Логистическая регрессия

In [18]:
best_model_lr_ros = best_model_lr(X_ros, y_ros)

Fitting 5 folds for each of 180 candidates, totalling 900 fits
Значение F1-score =  0.7058996011780503
Значение ROC-AUC =  0.7710281113538889
Лучшие параметры модели: {'C': 1, 'max_iter': 100, 'solver': 'newton-cg'}


Данный метод балансировки не самый лучший, так как дает отличные данные на тренировочных данных, модель переобучается, поэтому плохо работает на тестовых данных. Попробуем другие методы балансировки классов!

### SMOTE

In [19]:
X_smote, y_smote = SMOTE(random_state=123).fit_resample(X_train, y_train)

print("Original dataset:", sorted(Counter(y_train).items()))
print("SMOTE-sampled dataset:", sorted(Counter(y_ros).items()))

Original dataset: [(0, 5427), (1, 1391)]
SMOTE-sampled dataset: [(0, 5427), (1, 5427)]


#### Дерево решений

In [20]:
best_model_dtc_smote = best_model_dtc(X_smote, y_smote)

Fitting 5 folds for each of 1368 candidates, totalling 6840 fits
Значение F1-score =  0.8303317154575043
Значение ROC-AUC =  0.8550201050785374
Лучшие параметры модели: {'max_depth': 18, 'min_samples_leaf': 1, 'min_samples_split': 2}


#### Случайный лес

In [21]:
best_model_rfc_smote = best_model_rfc(X_smote, y_smote)

Fitting 5 folds for each of 1368 candidates, totalling 6840 fits
Fitting 5 folds for each of 33 candidates, totalling 165 fits
Значение F1-score =  0.9033760569588116
Значение ROC-AUC =  0.9654146610637555
Лучшие параметры модели: {'max_depth': 17, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 330}


#### Логистическая регрессия

In [23]:
best_model_lr_smote = best_model_lr(X_smote, y_smote)

Fitting 5 folds for each of 180 candidates, totalling 900 fits
Значение F1-score =  0.7274696958787252
Значение ROC-AUC =  0.8001325383964933
Лучшие параметры модели: {'C': 1, 'max_iter': 100, 'solver': 'sag'}


Данный метод балансировки лучше, чем RandomOverSampler, так как не дублирует существующие объекты, а добавляет новые со схожими показателями. Модель гораздо меньше переобучается и дает на тестовой выборке результат лучше.

### ADASYN 

In [24]:
X_adasyn, y_adasyn = ADASYN(random_state=123).fit_resample(X_train, y_train)

print("Original dataset:", sorted(Counter(y_train).items()))
print("SMOTE-sampled dataset:", sorted(Counter(y_adasyn).items()))

Original dataset: [(0, 5427), (1, 1391)]
SMOTE-sampled dataset: [(0, 5427), (1, 5193)]


#### Дерево решений

In [25]:
best_model_dtc_adasyn = best_model_dtc(X_adasyn, y_adasyn) 

Fitting 5 folds for each of 1368 candidates, totalling 6840 fits
Значение F1-score =  0.8059426527991509
Значение ROC-AUC =  0.8268292226709435
Лучшие параметры модели: {'max_depth': 19, 'min_samples_leaf': 1, 'min_samples_split': 2}


#### Случайный лес

In [26]:
best_model_rfc_adasyn = best_model_rfc(X_adasyn, y_adasyn) 

Fitting 5 folds for each of 1368 candidates, totalling 6840 fits
Fitting 5 folds for each of 33 candidates, totalling 165 fits
Значение F1-score =  0.8899499051554389
Значение ROC-AUC =  0.9577290694486591
Лучшие параметры модели: {'max_depth': 17, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 260}


#### Логистическая регрессия

In [27]:
best_model_lr_adasyn = best_model_lr(X_adasyn, y_adasyn)

Fitting 5 folds for each of 180 candidates, totalling 900 fits
Значение F1-score =  0.6948741234387951
Значение ROC-AUC =  0.764986584000353
Лучшие параметры модели: {'C': 1, 'max_iter': 100, 'solver': 'lbfgs'}


Данный метод балансировки лучше, чем RandomOverSampler. Добавляются новые обекты с показателями расположенными между 2 классами. Так модель начинает лучше понимать сложные случаи, где разделение на классы произвести достаточно сложно.

По итогам исследования наилучшие метрики набрала модель `RandomForest` при истользовании для балансировки классов метода **RandomOverSampler**: значение **F1-score = 0.957**, значение **ROC-AUC = 0.994**. Однако учитывая существующие недостатки данного метода семплирования (переобучение), в качестве наулучшей выбрана модель `RandomForest`, обученная на данных просемплированных методом **SMOTE**: значение **F1-score = 0.903**, значение **ROC-AUC = 0.965**. Проверим данную модель на тестовых данных!

## Тестирование модели

In [28]:
f1_score(y_test, best_model_rfc_smote.predict(X_test))

0.5991189427312776

In [33]:
roc_auc_score(y_test, best_model_rfc_smote.predict_proba(X_test)[:,1])

0.8485442048614011

## Выводы

Перед нами стояла задача разработки модели, которая могла бы прогнозировать уход клиента из банка в ближайшее время. Модель должна была иметь метрику качества **F1** не меньше 0.59.

В ходе проведенного исследования было выполнено:
- подготовка данных к моделированию (удалены пропуски в данных, категориальные данные закодированы, числовые данные масштабированы, выделены тренировочная и тестовая выборки);
- построены модели без учета баланса классов (DecisionTreeClassifier, RandomForestClassifier, LogisticRegression);
- применены различные методы балансировки данных (RandomOverSampler, SMOTE, ADASYN);
- построены модели с учетом баланса классов   

Наулучшей выбрана модель **RandomForest** с гиперпараметрами `max_depth` = 17, `min_samples_leaf` = 1, `min_samples_split` = 2, `n_estimators` = 330/ Обученние проходило на данных просемплированных методом **SMOTE**.   
На тестовой выборке значение **F1-score = 0.599**, значение **ROC-AUC = 0.849**.